[Serverless][Advanced Settings] Add validations on user input

ElenaStoeva opened the pull request:
Addresses https://github.com/elastic/kibana/issues/168684

## Summary

This PR adds validations of the user input for a some of the uiSettings (see https://github.com/elastic/kibana/issues/168684) based on the `schema` that is provided in the setting definitions. We do this by exposing a `validateValue` API from the uiSettings browser-side client, which calls an internal validate route to the server. Then `validateValue` is called in the `onChange` handlers of the setting input fields, utilising  `debounce` to reduce the number of requests to the server.

Note: One of the validation requirements is for validating that the `defaultIndex` setting value is an existing data view. This is a more complicated validation and will be addressed in a separate PR.

![defaultAnomalyScore](https://github.com/elastic/kibana/assets/59341489/4fb99993-9c2d-4c92-8ea4-bacdb58f7be4)
![defaultThreatIndex](https://github.com/elastic/kibana/assets/59341489/264fa663-43cb-44a1-9435-4de0fa1e957a)


### How to test
**Testing the validations in serverless:**
1. Start Es with `yarn es serverless` and Kibana with `yarn serverless-{es/oblt/security}`
2. Go to Management -> Advanced settings
3. Verify that the UI provides correct validation error messages for [the settings that require validation](https://github.com/elastic/kibana/issues/168684).

**Verify that self-managed Kibana is not affected:**
There should be no validations in self-managed Kibana and setting an invalid value for an advanced setting should be allowed and shouldn't cause any errors.

@elastic/kibana-core team, I know you're no longer code owners of the uiSettings service, but any feedback on those changes will be highly appreciated. 🙏 

<!---
### Checklist

Delete any items that are not applicable to this PR.

- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [ ] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [ ] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [ ] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [ ] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [ ] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)


### Risk Matrix

Delete this section if it is not applicable to this PR.

Before closing this PR, invite QA, stakeholders, and other developers to identify risks that should be tested prior to the change/feature release.

When forming the risk matrix, consider some of the following examples and how they may potentially impact the change:

| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|
| Multiple Spaces&mdash;unexpected behavior in non-default Kibana Space. | Low | High | Integration tests will verify that all features are still supported in non-default Kibana Space and when user switches between spaces. |
| Multiple nodes&mdash;Elasticsearch polling might have race conditions when multiple Kibana nodes are polling for the same tasks. | High | Low | Tasks are idempotent, so executing them multiple times will not result in logical error, but will degrade performance. To test for this case we add plenty of unit tests around this logic and document manual testing procedure. |
| Code should gracefully handle cases when feature X or plugin Y are disabled. | Medium | High | Unit tests will verify that any feature flag or plugin combination still results in our service operational. |
| [See more potential risk examples](https://github.com/elastic/kibana/blob/main/RISK_MATRIX.mdx) |


### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)
-->

ElenaStoeva committed:
[Serverless][Advanced Settings] Add validations

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/lint_ts_projects --fix'

ElenaStoeva committed:
Add internal route to server

ElenaStoeva committed:
Add config option for enabling the validations

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/lint_ts_projects --fix'

ElenaStoeva committed:
Refactor changes

ElenaStoeva committed:
Merge branch 'main' into serverless/advanced-settings/validations

ElenaStoeva committed:
Add tests

ElenaStoeva committed:
Update failing tests

ElenaStoeva committed:
Merge branch 'main' into serverless/advanced-settings/validations

ElenaStoeva committed:
Fix remaining type check errors

elasticmachine commented:
Pinging @elastic/platform-deployment-management (Team:Deployment Management)

kpollich reviewed (approved):
Fleet 🚀

drewdaemon reviewed (approved):
Visualizations changes look good.

davismcphee reviewed (approved):
Tested locally and confirmed `timepicker:quickRanges` and `defaultColumns` were validating as expected. Data Discovery changes LGTM 👍

ElenaStoeva committed:
Merge branch 'main' into serverless/advanced-settings/validations

ElenaStoeva committed:
Merge branch 'main' into serverless/advanced-settings/validations

jloleysens commented on src/core/server/integration_tests/ui_settings/doc_exists.ts:
```diff
@@ -201,6 +201,36 @@ export const docExistsSuite = (savedObjectsIndex: string) => () => {
   });
 
+  describe('validate route', () => {
+    it('returns a 200 and a validation error message if value is invalid', async () => {
```
Usually we return `400` for invalid request payloads. What's the reasoning behind using `200` here?

ElenaStoeva replied:
I was thinking that, since the purpose of this route is to validate a value and return a validation error message if one is thrown, we can accept any value and return `200` and an empty body if the value is valid, or `200` and a response body containing the validation error message if the value is invalid. WDYT?

jloleysens reviewed (commented):
Looks good, just left a q about `200` status code.

ElenaStoeva committed:
Merge branch 'main' into serverless/advanced-settings/validations

pgayvallet commented on packages/core/ui-settings/core-ui-settings-server/src/ui_settings_client.ts:
```diff
@@ -57,4 +57,8 @@ export interface IUiSettingsClient {
    * Shows whether the uiSetting is a sensitive value. Used by telemetry to not send sensitive values.
    */
   isSensitive: (key: string) => boolean;
+  /**
+   * Returns the validation error message, if one is thrown, when validating the uiSettings value.
+   */
+  getValidationErrorMessage: (key: string, value: unknown) => string | null;
```
NIT: maybe just `validate()`?

Also same remark as for the browser-side API, returning something more structured than `string | null` may be preferable.  

Also, to future-proof future usages (e.g async validations such as an index pattern), maybe returning a promise would be better as it would eventually allow later settings to perform async validation?

pgayvallet reviewed (commented):
Very happy to finally see server-side validation of the settings being added!

A few remarks / question on the ui-settings API changes 


EDIT: also a side question, I saw a single endpoint. How does that work between global and namespaced settings?

ElenaStoeva committed:
Address feedback

ElenaStoeva commented on packages/core/ui-settings/core-ui-settings-browser-internal/src/ui_settings_api.ts:
```diff
@@ -94,6 +98,31 @@ export class UiSettingsApi {
+        reject(error);
+      }
+    });
+  }
```
Yes, indeed this looks better. In fact, since we are anyway catching any errors in [validateValue()](https://github.com/elastic/kibana/blob/4f5fbcb2d5e14ab7f5fe3cfe6f0154bdaf436ca7/packages/core/ui-settings/core-ui-settings-browser-internal/src/ui_settings_client_common.ts#L138C7-L138C7), would it be wrong if we simply do it as below?
```ts
public async validate(key: string, value: any): Promise<ValidationApiResponse> {
    return await this.sendRequest('POST', `/internal/kibana/settings/${key}/validate`, {
      value,
    });
}
```

ElenaStoeva commented on packages/core/ui-settings/core-ui-settings-browser/src/types.ts:
```diff
@@ -100,6 +100,11 @@ export interface IUiSettingsClient {
+  /**
+   * Returns a validation error message if the uiSettings value is invalid.
+   */
+  validateValue: (key: string, value: any) => Promise<string | null>;
```
Yeah, that makes sense to me, especially for the server-side response, where it would be weird to return an empty response body when the value is valid. 

For the browser-side API, I thought it would be helpful to also return an additional boolean to indicate whether the validate request was successful:
```ts
type ValueValidation = { successfulValidation: boolean, valid?: boolean, errorMessage?: string };

validateValue: (key: string, value: any) => Promise<ValueValidation>;
```
The other option I considered was returning `Promise<ValueValidation | null>` so that we return `null` if we catch any errors, but I thought the pattern with `successfulValidation` might be more convenient to use on the client-side. WDYT?

ElenaStoeva commented on packages/core/ui-settings/core-ui-settings-server-internal/src/clients/base_ui_settings_client.ts:
```diff
@@ -72,6 +72,19 @@ export abstract class BaseUiSettingsClient implements IUiSettingsClient {
 
+  getValidationErrorMessage(key: string, value: unknown) {
+    const definition = this.defaults[key];
+    if (value === null || definition === undefined) return null;
```
Yeah, this logic doesn't look right.

In [4f5fbcb](https://github.com/elastic/kibana/pull/170234/commits/4f5fbcb2d5e14ab7f5fe3cfe6f0154bdaf436ca7),  I added throwing specific errors for these cases, based on which we return 404 for a setting key that doesn't exist, and 400 if the value is null. Does this make sense?

ElenaStoeva commented on packages/core/ui-settings/core-ui-settings-server-internal/src/routes/internal/validate.ts:
```diff
@@ -0,0 +1,71 @@
+  body: schema.object({
+    value: schema.any(),
+  }),
+};
```
Sure, that makes sense. Addressed with [4f5fbcb](https://github.com/elastic/kibana/pull/170234/commits/4f5fbcb2d5e14ab7f5fe3cfe6f0154bdaf436ca7)

ElenaStoeva commented on packages/core/ui-settings/core-ui-settings-server-internal/src/routes/internal/validate.ts:
```diff
@@ -0,0 +1,71 @@
+    response: KibanaResponseFactory
+  ) => {
+    try {
+      const { key } = request.params;
+      const { value } = request.body;
```
Mainly for consistency, since I followed the same pattern from the `set` route where the key is separated from the value.

ElenaStoeva commented:
Many thanks for the feedback @pgayvallet!

I addressed your comments and added some follow-up suggestions. Let me know what you think.


> EDIT: also a side question, I saw a single endpoint. How does that work between global and namespaced settings?

The validation requirement is for serverless only (to reduce performance issues), where there aren't any global settings in the UI, and as far as I know, there are no plans for adding validations in self-managed, which is why I didn't add an additional endpoint for global settings.

ElenaStoeva committed:
Fix failing tests

ElenaStoeva committed:
Merge branch 'main' into serverless/advanced-settings/validations

ElenaStoeva committed:
Merge branch 'main' into serverless/advanced-settings/validations

yuliacech reviewed (commented):
Great work on the settings validation, @ElenaStoeva! 
I've tested locally and noticed a couple things, let me know what you think:
1. Is it possible to display the validation constraints in the field description? For example, for "anomaly threshold" there is a note about valid values but nothing about how many indices are allowed in "threat indices" 
2. I was not able to find where the validation messages are coming from, is it something generated by the server automatically? I think in that case we won't be able to translate those message, which is a drawback for users with Kibana in a different language.
3. I noticed that the onBlur is only affecting some fields (probably only those that are being validated) and this creates some inconsistency in the UI. For example, if a not validated field is changed, the "save changes" bar is displayed instantaneously but if a validated field is changed, the user doesn't see any response from the UI at all unless they leave the field. This can be very confusing, I was not sure how to save a field when doing it for the 1st time. I also noticed that the json editor still validates with onChange and not onBlur, is this intended?

ElenaStoeva commented:
Thanks for the review @yuliacech!

> 1. Is it possible to display the validation constraints in the field description? For example, for "anomaly threshold" there is a note about valid values but nothing about how many indices are allowed in "threat indices"

The "anomaly threshold" setting has the constraints displayed in the description as this is part of the requirement for this setting (see https://github.com/elastic/kibana/issues/168684), but I don't see any requirements for displaying the limits in the description of the rest of the validated settings. It sounds reasonable to me to add them, but let's also see if @sixstringcode has any opinion on this.

> 2. I was not able to find where the validation messages are coming from, is it something generated by the server automatically? I think in that case we won't be able to translate those message, which is a drawback for users with Kibana in a different language.

Yes, the validation message is returned from the server which runs `schema.validate()` on the given value. Indeed, I also don't know how or whether we can translate them as they are dynamically generated. 😕 

> 3. I noticed that the onBlur is only affecting some fields (probably only those that are being validated) and this creates some inconsistency in the UI. For example, if a not validated field is changed, the "save changes" bar is displayed instantaneously but if a validated field is changed, the user doesn't see any response from the UI at all unless they leave the field. This can be very confusing, I was not sure how to save a field when doing it for the 1st time. I also noticed that the json editor still validates with onChange and not onBlur, is this intended?

My idea was to switch to `onBlur` only for the fields that have validation - array (which already used `onBlur` btw), number, and the code editor field, to reduce performance penalty from the server requests, and I thought it would be okay to leave `onChange` for the rest of the fields since IMO it creates a better UX as it provides immediate feedback from the UI. However, the code editor field type uses a `CodeEditor` component, which doesn't seem to support `onBlur`, so I don't see how we can switch it to `onBlur`. It's also a good point that using `onBlur` for some fields and `onChange` for others creates some inconsistency in the UI. We could switch all fields to `onBlur` but this might be wasteful if, at the end, the code editor field still uses `onChange` and there will still be inconsistency between this field and the rest.

pgayvallet reviewed (approved):
Thanks for the changes, LGTM

janmonschke reviewed (approved):
Threat Hunting Investigations changes lgtm 🎉

yuliacech commented:
Thanks for answering my questions, @ElenaStoeva! 1. and 2. are non-blockers for me for this PR, so please feel free to follow up as needed or open corresponding issues. For 3. I think onBlur introduces not only inconsistency but a very confusing UX when the user changes the value but no "save" button is being displayed. I would suggest using onChange as before but add a "debounce" to the validate call: the request is only sent after the user stopped typing for some time.

ElenaStoeva committed:
Switch to onChange with debounce

ElenaStoeva committed:
Fix field input jest tests

ElenaStoeva committed:
Merge branch 'main' into serverless/advanced-settings/validations

ElenaStoeva commented:
Thank you so much for the review @yuliacech! I think using `onChange` with `debounce` is a great idea! I implemented this in [ba18bd1](https://github.com/elastic/kibana/pull/170234/commits/ba18bd1366962d4a7d57ab7175e45c8b8b5f54d2) - please let me know what you think.

I also opened https://github.com/elastic/kibana/issues/171382 to track the work for displaying the limits in the descriptions.

yuliacech reviewed (approved):
Thanks a lot for addressing my comments, @ElenaStoeva! 
Latest changes LGTM 🚀

clintandrewhall reviewed (approved):
This has been extensively reviewed, and the changes to the tests LGTM.  As long as the tests pass, I'm glad to see this merged!

ElenaStoeva committed:
Merge branch 'main' into serverless/advanced-settings/validations
